### PR TITLE
Add resources for HPA.

### PR DESCRIPTION
**Pull request by MarvinCai**
more detail in readme

### CODE REVIEW

File: modules/hpa/main.tf [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+  }
+}
+
+resource "kubernetes_manifest" "serviceaccount_sn_system_scaling_prometheus" {
+  manifest = {
+    "apiVersion" = "v1"
+    "kind" = "ServiceAccount"
+    "metadata" = {
+      "annotations" = {
+        "iam.gke.io/gcp-service-account" = "pulsar-prometheus@sncloud-dev-zxc.iam.gserviceaccount.com"
```
Comment (jrsdav) [inline modules/hpa/main.tf:96]: The value for this annotation should be a variable that you pass to the module

File: modules/hpa/main.tf [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+      }
+      "name" = "scaling-prometheus"
+      "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:99]: The namespace should also be a value passed to the module

File: modules/hpa/main.tf [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+    "roleRef" = {
+      "apiGroup" = "rbac.authorization.k8s.io"
+      "kind" = "ClusterRole"
+      "name" = "scaling-prometheus"
+    }
+    "subjects" = [
+      {
+        "kind" = "ServiceAccount"
+        "name" = "scaling-prometheus"
+        "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:120]: See earlier comment

File: modules/hpa/main.tf [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+  }
+}
+
+resource "kubernetes_manifest" "prometheus_sn_system_scaling_prometheus" {
+  manifest = {
+    "apiVersion" = "monitoring.coreos.com/v1"
+    "kind" = "Prometheus"
+    "metadata" = {
+      "name" = "scaling-prometheus"
+      "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:203]: See earlier comment

File: modules/hpa/main.tf [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+  }
+}
+
+resource "kubernetes_manifest" "service_sn_system_scaling_prometheus" {
+  manifest = {
+    "apiVersion" = "v1"
+    "kind" = "Service"
+    "metadata" = {
+      "name" = "scaling-prometheus"
+      "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:249]: See earlier comment

File: modules/hpa/main.tf [hunk 6]
```diff
@@ -0,0 +1,757 @@
...
+}
+
+resource "kubernetes_manifest" "secret_issuer_cert" {
+  computed_fields = ["stringData"]
+  manifest = {
+    "apiVersion" = "v1"
+    "kind" = "Secret"
+    "metadata" = {
+      "name" = "issuer-cert"
+      "namespace" = "cert-manager"
```
Comment (jrsdav) [inline modules/hpa/main.tf:303]: We're not currently using this namespace. Should we be putting cert-manager into its own dedicated namespace?

File: modules/hpa/main.tf [hunk 7]
```diff
@@ -0,0 +1,757 @@
...
+}
+
+# cert for kube-aggregator to use for communicating with prometheus adapter
+resource "kubernetes_manifest" "certificate_sn_system_custom_metrics_server" {
+  manifest = {
+    "apiVersion" = "cert-manager.io/v1"
+    "kind" = "Certificate"
+    "metadata" = {
+      "name" = "custom-metrics-server"
+      "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:344]: See earlier comments

File: modules/hpa/main.tf [hunk 8]
```diff
@@ -0,0 +1,757 @@
...
+  ]
+}
+
+resource "kubernetes_manifest" "serviceaccount_sn_system_custom_metrics_apiserver" {
+  manifest = {
+    "apiVersion" = "v1"
+    "kind" = "ServiceAccount"
+    "metadata" = {
+      "name" = "custom-metrics-apiserver"
+      "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:372]: See earlier comments

File: modules/hpa/main.tf [hunk 9]
```diff
@@ -0,0 +1,757 @@
...
+    "roleRef" = {
+      "apiGroup" = "rbac.authorization.k8s.io"
+      "kind" = "ClusterRole"
+      "name" = "system:auth-delegator"
+    }
+    "subjects" = [
+      {
+        "kind" = "ServiceAccount"
+        "name" = "custom-metrics-apiserver"
+        "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:393]: See earlier comments

File: modules/hpa/main.tf [hunk 10]
```diff
@@ -0,0 +1,757 @@
...
+    "roleRef" = {
+      "apiGroup" = "rbac.authorization.k8s.io"
+      "kind" = "ClusterRole"
+      "name" = "custom-metrics-resource-reader"
+    }
+    "subjects" = [
+      {
+        "kind" = "ServiceAccount"
+        "name" = "custom-metrics-apiserver"
+        "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:511]: See earlier comments

File: modules/hpa/main.tf [hunk 11]
```diff
@@ -0,0 +1,757 @@
...
+resource "kubernetes_manifest" "deployment_sn_system_custom_metrics_apiserver" {
+  manifest = {
+    "apiVersion" = "apps/v1"
+    "kind" = "Deployment"
+    "metadata" = {
+      "labels" = {
+        "app" = "custom-metrics-apiserver"
+      }
+      "name" = "custom-metrics-apiserver"
+      "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:526]: See earlier comments

File: modules/hpa/main.tf [hunk 12]
```diff
@@ -0,0 +1,757 @@
...
+  }
+}
+
+resource "kubernetes_manifest" "service_sn_system_custom_metrics_apiserver" {
+  manifest = {
+    "apiVersion" = "v1"
+    "kind" = "Service"
+    "metadata" = {
+      "name" = "custom-metrics-apiserver"
+      "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:622]: See earlier comments

File: modules/hpa/main.tf [hunk 13]
```diff
@@ -0,0 +1,757 @@
...
+    "metadata" = {
+      "name" = "v1beta1.custom.metrics.k8s.io"
+    }
+    "spec" = {
+      "group" = "custom.metrics.k8s.io"
+      "groupPriorityMinimum" = 100
+      "insecureSkipTLSVerify" = true
+      "service" = {
+        "name" = "custom-metrics-apiserver"
+        "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:652]: See earlier comments

File: modules/hpa/main.tf [hunk 14]
```diff
@@ -0,0 +1,757 @@
...
+    "metadata" = {
+      "name" = "v1beta2.custom.metrics.k8s.io"
+    }
+    "spec" = {
+      "group" = "custom.metrics.k8s.io"
+      "groupPriorityMinimum" = 100
+      "insecureSkipTLSVerify" = true
+      "service" = {
+        "name" = "custom-metrics-apiserver"
+        "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:673]: See earlier comments

File: modules/hpa/main.tf [hunk 15]
```diff
@@ -0,0 +1,757 @@
...
+          matches: 'container_fs_writes_bytes_total'
+          as: 'disk_write_rate_kb'
+        metricsQuery: sum(rate(container_fs_writes_bytes_total{<<.LabelMatchers>>}[5m])) by (<<.GroupBy>>) / 1024
+
+      EOT
+    }
+    "kind" = "ConfigMap"
+    "metadata" = {
+      "name" = "adapter-config"
+      "namespace" = "sn-system"
```
Comment (jrsdav) [inline modules/hpa/main.tf:754]: See earlier comments

**Review by jrsdav — COMMENTED** [review]
Looks good, other than a few hard coded values should be input variables. This just allows for future flexibility if we need to configure in a non-standard environment

File: modules/hpa/main.tf (hunk 1 above)
Comment (MarvinCai) [reply modules/hpa/main.tf:96]: seems added when I tested in my dev env while I've try deploy this in test env GCP and AWS cluster and works fine, so this should be needed.

File: modules/hpa/main.tf (hunk 6 above)
Comment (MarvinCai) [reply modules/hpa/main.tf:303]: I see we're using sn's namespace for cert-manager, just added a dedicated var for it.

**Comment by MarvinCai** [comment]
I made all namespace configurable and created 3 variables, one to specify namespace for the prometheus, one to specify where cert-manager is installed, one to specify namespace for metric server installation.

**Review by jrsdav — APPROVED** [review]
LGTM!

**Pull request merged**